### ISSUE TITLE
Support Output/Input stream for protobuf format 

### ISSUE THREAD

**Issue by GeorgePap-719** [Kotlin/kotlinx.serialization#2075]
**What is your use-case and why do you need this feature?**

I'm using springboot and RSocket APIs with protobuf messages ( using kotlinx.serialization impl) and currently to support it i'm using my own custom spring-kotlin-encoder "glue-code". I would like to provide the first step to make it work out of the box ( i guess ktor also could use it?)  for simple and  [delimited messages](https://developers.google.com/protocol-buffers/docs/techniques?hl=en#streaming) 

Imho: I think the first step would be to provide jvm encoders that take input as io-stream.  

**Describe the solution you'd like**

Probably similar APIs like json format has:

`ProtoBuf.encodeToStream`

`ProtoBuf.decodeFromStream`

And for delimited messages:

`ProtoBuf.encodeDelimitedMessagesToStream`

`ProtoBuf.decodeDelimitedMessagesFromStream`

**Comment by bishiboosh** [comment]
I would be quite interested by this to be able to use it Android Datastore. Can we take inspiration from what's been done in [Wire](https://github.com/square/wire) maybe ?

**Comment by GeorgePap-719** [comment]
Ofc, feel free to take anything you like.

**Comment by ShreckYe** [comment]
A vote for this.

One question: why don't you open a pull request or draft pull request for this? it can make reviewing changes easier.

And an immature suggestion: can we get it all multiplatform with [Coroutines channels](https://kotlinlang.org/docs/channels.html) or [kotlinx-io](https://github.com/Kotlin/kotlinx-io)? With channels we get suspending/non-blocking streaming support too. kotlinx-io provides the `Sink` and `Source` APIs though I am not sure how stable it is.

**Comment by GeorgePap-719** [comment]
Hey @ShreckYe, what is your use-case for this feature? Can you explain it here, please.

> can we get it all multiplatform with [Coroutines channels](https://kotlinlang.org/docs/channels.html) or [kotlinx-io](https://github.com/Kotlin/kotlinx-io)? With channels we get suspending/non-blocking streaming support too. kotlinx-io provides the Sink and Source APIs though I am not sure how stable it is.

Usually these kind of low-level APIs are blocking for performance reasons. However, I can provide APIs which return a  `Sequence`, similar to what `Json` format provides.

**Comment by ShreckYe** [comment]
Hi @GeorgePap-719, thanks for asking. I mainly care about **multiplatform** support now. Suspending support can be an extra bonus but I don't need it urgently.

> Usually these kind of low-level APIs are blocking for performance reasons.

In my opinion this is half valid. If you are only encoding and decoding small messages this can be true, which does happen in most use cases, because the bytes are probably written and read into buffers and one encoding or decoding operation might encode or decode the message as a whole to or from the buffer, not reaching the end involving any IO waiting for the buffer to be reset. However for large messages, these operations will involve IO and waiting. And under asynchronous/non-blocking/reactive architecture, such as Vert.x in my case, reactive/suspending IO does improve throughput.

It's true that a major obstacle here is that if these decoding/encoding functions are made `suspend`, the supporting atomic operations and thus the whole `Encoder`/`Decoder` hierarchy will also be made `suspend`, which requires a lot of code to be refactored and I suppose you are already aware of. Kotlin Coroutines is based on CPS transformation, so there is also some overhead. So does improvement brought by reactiveness/suspending justify the sacrifice in overhead in terms of throughput? I don't know. Maybe only implementing and benchmarking it can tell. However, it's worth noting that Netty has implemented [ProtobufEncoder](https://netty.io/4.0/api/io/netty/handler/codec/protobuf/ProtobufEncoder.html) and [ProtobufDecoder](https://netty.io/4.1/api/io/netty/handler/codec/protobuf/ProtobufDecoder.html).

I think another advantage of this is that if we support this we also start supporting network streaming directly using a serialization format. To think further (immaturely), if we support `Flow`s/`Sequence`s in kotlinx.serialization, giving them the same status as `List`s, we can just send and receive `Flow`s and get a kind of a streaming protocol out of the box with serialization.

**Comment by ShreckYe** [comment]
On second thought about this, I would say that my main concern is multiplatform support and Coroutines support is not that important at all. Serializing large messages is a rare use case anyway. And it's still arguable that even for a large message in a suspending context, is it faster to serialize it to a `ByteArray` first and then send it, or to wrap the sending operation in a `SendChannel` and pass it to the "suspend serialize function" which depends on smaller suspend ones? And since throughput is mainly a server-side concern, with JVM Loom / virtual threads gradually becoming stable and Coroutines integrating it, we can just run the blocking APIs on virtual threads and easily get a rough estimate of performance.

So to sum up, I think it would be nice to use the `Source` and `Sink` APIs from kotlinx-io.

**Comment by GeorgePap-719** [comment]
This probably could also be used by KTOR to support grpc, see: [KTOR-1501](https://youtrack.jetbrains.com/issue/KTOR-1501) for more details.

My initial POC targeted only the jvm. To make it multiplatform, we would have to wait for `kotlinx-io`. On the other hand, Json-format supports streaming only for the jvm.
